src/main/java/net/floodlightcontroller/forwarding/Forwarding.java

```diff
@@ -228,46 +200,661 @@ private void removeExpiredFlowSetId(U64 flowSetId, NodePortTuple avoid, Iterator
+     * @return
+     */
+    private boolean isBroadcastOrMulticast(Ethernet eth) {
+        return eth.isBroadcast() || eth.isMulticast();
+    }
+
+    /**
+     * L3 routing block -- it supports routing across subnet by using "virtual gateway". It also keeps L2 forwarding
+     * as Floodlight original does
+     *
+     * @param eth Ethernet Frame that associated with this OFPacketIn
+     * @param sw Switch on which the packet was received
+     * @param pi FPacketIn that came to the switch
+     * @param decision Decision that caused flooding, or null
+     * @param cntx FloodlightContext associated with this OFPacketIn
+     * @param gatewayInstance Virtual gateway instance
+     * @param inPort Ingress switch port on which the packet was received
+     */
+    protected void doL3Routing(Ethernet eth, IOFSwitch sw, OFPacketIn pi, IRoutingDecision decision,
+                             FloodlightContext cntx, @Nonnull VirtualGatewayInstance gatewayInstance, OFPort inPort) {
+
+        MacAddress gatewayMac = gatewayInstance.getGatewayMac();
+
+        if (eth.getPayload() instanceof IPv4) {
+            IPv4Address intfIpAddress = findInterfaceIP(gatewayInstance, ((IPv4) eth.getPayload()).getDestinationAddress());
+            if (intfIpAddress == null) {
+                log.debug("Can not locate corresponding interface for gateway {}, check its interface configuration",
+                        gatewayInstance.getName());
+                return;
+            }
+        }
+
+        if (isBroadcastOrMulticast(eth)) {
+            // When cross-subnet, host send ARP request to gateway. Gateway need to generate ARP response to host
+            if (eth.getPayload() instanceof ARP && ((ARP) eth.getPayload()).getOpCode().equals(ARP.OP_REQUEST)
+                    && gatewayInstance.isAGatewayInft(((ARP) eth.getPayload()).getTargetProtocolAddress())) {
+                IPacket arpReply = gatewayArpReply(cntx, gatewayMac);
+                pushArpReply(arpReply, sw, OFBufferId.NO_BUFFER, OFPort.ANY, inPort);
+                log.debug("Virtual gateway pushing ARP reply message to source host");
+            }
+            else {
+                doFlood(sw, pi, decision, cntx);
+            }
+        }
+        else {
+            // This also includes L2 forwarding
+            doL3ForwardFlow(sw, pi, decision, cntx, gatewayInstance, false);
+        }
+
+    }
+
+
+    /**
+     * This function retrieves a patch from source device to destination device and then install flows(L3/L2) over the path.
+     * In L3 case, virtual router caches the packet-in while it sends the ARP message to learn destination device
+     *
+     * @param sw
+     * @param pi
+     * @param decision
+     * @param cntx
+     * @param gateway
+     * @param requestFlowRemovedNotifn
+     */
+    protected void doL3ForwardFlow(IOFSwitch sw, OFPacketIn pi, IRoutingDecision decision, FloodlightContext cntx,
```

Comment:
This function is pretty big. It's definitely not critical, but it would probably be good for readability to break this up